src/tir/schedule/primitive/reduction.cc

```diff
@@ -284,7 +284,22 @@ StmtSRef DecomposeReduction(ScheduleState self, const StmtSRef& block_sref,
                /*body=*/body);
   }
   body = Substitute(body, loop_var_map);
-  // Step 6. Mutate IR
+  // Step 6. Add write regions back to read regions in update block.
+  Array<BufferRegion> new_reads;
+  std::unordered_set<const BufferNode*> read_bufs;
+  for (const BufferRegion& read_access : block->reads) {
+    read_bufs.insert(read_access->buffer.get());
+  }
+  for (const BufferRegion& write_access : block->writes) {
+    if (read_bufs.find(write_access->buffer.get()) == read_bufs.end()) {
+      new_reads.push_back(write_access);
+    }
+  }
+  for (const BufferRegion& read_access : block->reads) {
+    new_reads.push_back(read_access);
+  }
+  (const_cast<BlockNode*>(block))->reads = std::move(new_reads);
```

Comment:
After this PR, I'm getting a structural hash mismatch in meta schedule `ApplyHistoryBest` https://github.com/apache/tvm/blob/ce335c3a74185df6cc1152e53c60695d8a418d8e/src/meta_schedule/integration.cc#L134

I found that if I do `decompose_reduction`, IRModules that have been committed to the database are modified. Specifically, the query module has
```
reads([placeholder[b, i, k], T_layout_trans[b, floordiv(j, 16), floordiv(k, 4), floormod(j, 16), floormod(k, 4)]])
writes([compute[b, i, j]])
...
with init() {
  compute[b, i, j] = 0
}
compute[b, i, j] = (compute[b, i, j] + (int32(placeholder[b, i, k])*int32(T_layout_trans[b, floordiv(j, 16), floordiv(k, 4), floormod(j, 16), floormod(k, 4)])))
```
but the corresponding mod in the database has different `reads`:
```
reads(compute[b, i, j], [placeholder[b, i, k], T_layout_trans[b, floordiv(j, 16), floordiv(k, 4), floormod(j, 16), floormod(k, 4)]])
...
```

I haven't looked at what this PR does, but I'm pretty sure this line is the offending bug...

Reply:
@masahi thanks for reporting this! I'm not sure why the IRModule is mutated in place after decompose-reduction (not sure if i understand correctly), but would be great to have a minimal reproducible example after you returning from vacation :-)

@yzh119 would you mind taking a look at the particular case Masa mentions and perhaps add a regression test?

Reply:
yeah, I can prepare a repro. It's a bit complicated since it involves scheduling, database stuff, and `ApplyHistoryBest`. I know the issue comes from this PR since I bisected it.

So when `decompose_reduction` is called, we are now executing

https://github.com/yzh119/tvm/blob/cd42ab101ab993edae2956fdc0415561be5f9718/src/tir/schedule/primitive/reduction.cc#L301

and I'm guess that this in-place update propagates all the way up to the committed `prim_mod` in the tuning database.

Reply:
ok the repro is ready: https://gist.github.com/masahi/591078723d26f09ece3430af95835c99

The test works with the current `main`. When I run it, I get this output:
```
With decompose_reduction=True
One or more operators have not been tuned. Please tune your model for better performance. Use DEBUG logging level to see more details.
[07:20:13] /home/masa/projects/dev/tvm/src/meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_nn_contrib_dense_pack
[07:20:13] /home/masa/projects/dev/tvm/src/meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_expand_dims
[07:20:13] /home/masa/projects/dev/tvm/src/meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_cast
[07:20:13] /home/masa/projects/dev/tvm/src/meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_cast_1
[07:20:13] /home/masa/projects/dev/tvm/src/meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_nn_batch_matmul

With decompose_reduction=False
[07:20:13] /home/masa/projects/dev/tvm/src/meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_expand_dims
[07:20:13] /home/masa/projects/dev/tvm/src/meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_cast
[07:20:13] /home/masa/projects/dev/tvm/src/meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_cast_1
```

This means, if `decompose_reduction` is applied during scheduling, `ApplyHistoryBest` fails to match the structural hash of the query mod, corresponding to `nn_contrib_dense` and `nn_batch_matmul`, against the ones in database.

If I revert this commit, I get the following expected output
```
With decompose_reduction=True
One or more operators have not been tuned. Please tune your model for better performance. Use DEBUG logging level to see more details.
[07:24:21] /home/masa/projects/dev/tvm/src/meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_expand_dims
[07:24:21] /home/masa/projects/dev/tvm/src/meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_cast
[07:24:21] /home/masa/projects/dev/tvm/src/meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_cast_1

With decompose_reduction=False
[07:24:21] /home/masa/projects/dev/tvm/src/meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_expand_dims
[07:24:21] /home/masa/projects/dev/tvm/src/meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_cast
[07:24:21] /home/masa/projects/dev/tvm/src/meta_schedule/integration.cc:146: Warning: Cannot find workload: tvmgen_default_fused_cast_1
```

You can dump the contents of the database at https://github.com/apache/tvm/blob/b08e8e49b5ea0e406c3ca41610a131dd271f0d41/src/meta_schedule/database/json_database.cc#L72-L74 to see that `reads` region for dense and bmm is modified if `decompose_reduction` is enabled.